core/grant/service.go

```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
 	return newAndUpdatedGrants, nil
 }
 
+func (s *Service) DormancyCheck(ctx context.Context, criteria domain.DormancyCheckCriteria) error {
+	if err := criteria.Validate(); err != nil {
+		return fmt.Errorf("invalid criteria: %w", err)
+	}
+	startDate := time.Now().Add(-criteria.Period)
+
+	provider, err := s.providerService.GetByID(ctx, criteria.ProviderID)
+	if err != nil {
+		return fmt.Errorf("getting provider details: %w", err)
+	}
+
+	s.logger.Info(fmt.Sprintf("getting active grants for provider %q", provider.URN))
+	grants, err := s.List(ctx, domain.ListGrantsFilter{
+		Statuses:      []string{string(domain.GrantStatusActive)}, // TODO: evaluate later to use status_in_provider
+		ProviderTypes: []string{provider.Type},
+		ProviderURNs:  []string{provider.URN},
+		CreatedAtLte:  startDate,
+	})
+	if err != nil {
+		return fmt.Errorf("listing active grants: %w", err)
+	}
+	if len(grants) == 0 {
+		s.logger.Info(fmt.Sprintf("no active grants found for provider %q", provider.URN))
+		return nil
+	}
+	grantIDs := getGrantIDs(grants)
+	s.logger.Info(fmt.Sprintf("found %d active grants for provider %q", len(grants), provider.URN), "grant_ids", grantIDs)
+
+	var uniqueAccountIDs []string
+	uniqueAccountIDsMap := map[string]bool{}
+	for _, g := range grants {
+		if _, ok := uniqueAccountIDsMap[g.AccountID]; !ok {
+			uniqueAccountIDs = append(uniqueAccountIDs, g.AccountID)
+			uniqueAccountIDsMap[g.AccountID] = true
+		}
+	}
+
+	s.logger.Info(fmt.Sprintf("getting activities for provider %q", provider.URN))
+	activities, err := s.providerService.ListActivities(ctx, *provider, domain.ListActivitiesFilter{
+		AccountIDs:   uniqueAccountIDs,
+		TimestampGte: &startDate,
+	})
+	if err != nil {
+		return fmt.Errorf("listing activities: %w", err)
+	}
+
+	grantsPointer := make([]*domain.Grant, len(grants))
+	for i, g := range grants {
+		g := g
+		grantsPointer[i] = &g
+	}
+	if err := s.providerService.CorrelateGrantActivities(ctx, *provider, grantsPointer, activities); err != nil {
+		return fmt.Errorf("correlating grant activities: %w", err)
+	}
+
+	s.logger.Info("checking grants dormancy...")
+	var dormantGrants []*domain.Grant
+	var dormantGrantsIDs []string
+	var dormantGrantsByOwner = map[string][]*domain.Grant{}
+	for _, g := range grantsPointer {
+		if len(g.Activities) == 0 {
+			g.ExpirationDateReason = fmt.Sprintf("%s: %s", domain.GrantExpirationReasonDormant, criteria.RetainDuration)
+			newExpDate := time.Now().Add(criteria.RetainDuration)
+			g.ExpirationDate = &newExpDate
```

Comment:
Can we keep this expiration date intact as it is as its based on user input, and create another column explicitly for this functionality? 
and in the grant expiration reminder and revocation job, the new field can be included for expiry.

Reply:
we have introduced `requested_expiration_date` in #27 which will record the initial expiration date as requested in the appeal and this field will never be updated.